### PR TITLE
C#: Detect constantly failing assertions in the CFG

### PR DESCRIPTION
**Pull request by hvitved**
In addition to detecting failing assertions such as `Debug.Assert(false)`, this PR also introduces _exit_ completions: An exit completion is used to properly model calls to e.g. `System.Environment.Exit()`, which exits an application directly, skipping potential `finally` blocks.

Commit-by-commit review is suggested.

@calumgrant

### CODE REVIEW

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll [hunk 1]
```diff
@@ -1774,6 +1775,11 @@ module ControlFlow {
         private class ExitingCall extends NonReturningCall {
           ExitingCall() {
             this.getTarget() instanceof ExitingCallable
+            or
+            exists(AssertMethod m |
+              m = this.(FailingAssertion).getAssertMethod() |
+              not exists(m.getExceptionClass())
```
Comment (calumgrant) [inline csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: Are there actually any Assert methods that do terminate the process when they fail?

File: change-notes/1.19/analysis-csharp.md [hunk 2]
```diff
@@ -2,8 +2,10 @@
 
 ## General improvements
 
-* The control flow graph construction now takes simple Boolean conditions on local scope variables into account. For example, in `if (b) x = 0; if (b) x = 1;`, the control flow graph will reflect that taking the `true` (resp. `false`) branch in the first condition implies taking the same branch in the second condition. In effect, the first assignment to `x` will now be identified as being dead.
-
+* Control flow graph improvements:
+  * The control flow graph construction now takes simple Boolean conditions on local scope variables into account. For example, in `if (b) x = 0; if (b) x = 1;`, the control flow graph will reflect that taking the `true` (resp. `false`) branch in the first condition implies taking the same branch in the second condition. In effect, the first assignment to `x` will now be identified as being dead.
+  * Constant failing assertions, such as `Debug.Assert(false)`, are now taken into account. Syntactic successors of constant failing assertions are consequently dead code.
```
Comment (calumgrant) [inline change-notes/1.19/analysis-csharp.md]: How about
```
Code that is only reachable from a constant failing assertion is considered to be unreachable.
```

File: csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll (hunk 1 above)
Comment (hvitved) [reply csharp/ql/src/semmle/code/csharp/controlflow/ControlFlowGraph.qll]: Debug.Assert

**Comment by jf205** [comment]
@hvitved change note LGTM

**Pull request merged**